backend/cfg/cfg_regalloc_utils.ml

```diff
@@ -450,3 +455,83 @@ let may_use_stack_operands_everywhere :
   may_use_stack_operands_array map instr.arg;
   may_use_stack_operands_array map instr.res;
   All_spilled_registers_rewritten
+
+let insert_block :
+    Cfg_with_layout.t ->
+    Cfg.BasicInstructionList.t ->
+    after:Cfg.basic_block ->
+    next_instruction_id:(unit -> Instruction.id) ->
+    unit =
+ fun cfg_with_layout body ~after:predecessor_block ~next_instruction_id ->
+  let cfg = Cfg_with_layout.cfg cfg_with_layout in
+  let successors =
+    Cfg.successor_labels ~normal:true ~exn:false predecessor_block
+  in
+  (* CR gyorsh: this function is currently only when |successors|>=1 and |body|
+     > 0*)
+  if Label.Set.cardinal successors = 0
+  then
+    Misc.fatal_errorf
+      "Cannot insert a block after block %a: it has no successors" Label.print
+      predecessor_block.start;
+  let last_insn =
+    match Cfg.BasicInstructionList.last body with
+    | None -> Misc.fatal_error "Inserting an empty block"
+    | Some i -> i
+  in
+  let copy (i : Cfg.basic Cfg.instruction) : Cfg.basic Cfg.instruction =
+    { i with id = next_instruction_id () }
+  in
+  (* copy body if there is more than one successor *)
+  let first = ref true in
+  let get_body () =
+    if !first
+    then (
+      first := false;
+      body)
+    else
+      let new_body = Cfg.BasicInstructionList.make_empty () in
+      Cfg.BasicInstructionList.iter body ~f:(fun instr ->
+          Cfg.BasicInstructionList.add_end new_body (copy instr));
+      new_body
+  in
+  Label.Set.iter
+    (fun successor_label ->
+      let successor_block = Cfg.get_block_exn cfg successor_label in
+      let start = Cmm.new_label () in
+      let block : Cfg.basic_block =
+        { start;
+          body = get_body ();
+          terminator =
+            { (* The [successor_block] is the only successor. *)
```

Comment:
```suggestion
            {
```
presumably no longer true

Reply:
It is true, the `successor_block` is the only successor of this new block.

In this implementation we add new blocks for every successor of original block.